In [1]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix,roc_auc_score,roc_curve,accuracy_score, classification_report
from sklearn import metrics
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')


## Data Cleaning

In [2]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/00468/online_shoppers_intention.csv"
df = pd.read_csv(url)
df.head()

Administrative  Administrative_Duration  Informational  \
0               0                      0.0              0   
1               0                      0.0              0   
2               0                      0.0              0   
3               0                      0.0              0   
4               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0               1                 0.000000   
1                     0.0               2                64.000000   
2                     0.0               1                 0.000000   
3                     0.0               2                 2.666667   
4                     0.0              10               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.20       0.20         0.0         0.0   Feb                 1   
1         0.00       0.10         0.0         0.0   Feb                 2   
2         0.20       0.20         0.0         0.0   Feb                 4   
3         0.05       0.14         0.0         0.0   Feb                 3   
4         0.02       0.05         0.0         0.0   Feb                 3   

   Browser  Region  TrafficType        VisitorType  Weekend  Revenue  
0        1       1            1  Returning_Visitor    False    False  
1        2       1            2  Returning_Visitor    False    False  
2        1       9            3  Returning_Visitor    False    False  
3        2       2            4  Returning_Visitor    False    False  
4        3       1            4  Returning_Visitor     True    False

In [3]:
print("The data has {} data points".format(df.shape[0]))
print("The data has {} features".format(df.shape[1]))

The data has 12330 data points
The data has 18 features


In [4]:
# Drop the'Browser' column
df.drop(columns = ['Browser'], inplace=True)
df.head()

Administrative  Administrative_Duration  Informational  \
0               0                      0.0              0   
1               0                      0.0              0   
2               0                      0.0              0   
3               0                      0.0              0   
4               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0               1                 0.000000   
1                     0.0               2                64.000000   
2                     0.0               1                 0.000000   
3                     0.0               2                 2.666667   
4                     0.0              10               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.20       0.20         0.0         0.0   Feb                 1   
1         0.00       0.10         0.0         0.0   Feb                 2   
2         0.20       0.20         0.0         0.0   Feb                 4   
3         0.05       0.14         0.0         0.0   Feb                 3   
4         0.02       0.05         0.0         0.0   Feb                 3   

   Region  TrafficType        VisitorType  Weekend  Revenue  
0       1            1  Returning_Visitor    False    False  
1       1            2  Returning_Visitor    False    False  
2       9            3  Returning_Visitor    False    False  
3       2            4  Returning_Visitor    False    False  
4       1            4  Returning_Visitor     True    False

In [5]:
print("The data has {} data points".format(df.shape[0]))
print("The data has {} features".format(df.shape[1]))

The data has 12330 data points
The data has 17 features


In [6]:
(df.describe())

Administrative  Administrative_Duration  Informational  \
count    12330.000000             12330.000000   12330.000000   
mean         2.315166                80.818611       0.503569   
std          3.321784               176.779107       1.270156   
min          0.000000                 0.000000       0.000000   
25%          0.000000                 0.000000       0.000000   
50%          1.000000                 7.500000       0.000000   
75%          4.000000                93.256250       0.000000   
max         27.000000              3398.750000      24.000000   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
count            12330.000000    12330.000000             12330.000000   
mean                34.472398       31.731468              1194.746220   
std                140.749294       44.475503              1913.669288   
min                  0.000000        0.000000                 0.000000   
25%                  0.000000        7.000000               184.137500   
50%                  0.000000       18.000000               598.936905   
75%                  0.000000       38.000000              1464.157214   
max               2549.375000      705.000000             63973.522230   

        BounceRates     ExitRates    PageValues    SpecialDay  \
count  12330.000000  12330.000000  12330.000000  12330.000000   
mean       0.022191      0.043073      5.889258      0.061427   
std        0.048488      0.048597     18.568437      0.198917   
min        0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.014286      0.000000      0.000000   
50%        0.003112      0.025156      0.000000      0.000000   
75%        0.016813      0.050000      0.000000      0.000000   
max        0.200000      0.200000    361.763742      1.000000   

       OperatingSystems        Region   TrafficType  
count      12330.000000  12330.000000  12330.000000  
mean           2.124006      3.147364      4.069586  
std            0.911325      2.401591      4.025169  
min            1.000000      1.000000      1.000000  
25%            2.000000      1.000000      2.000000  
50%            2.000000      3.000000      2.000000  
75%            3.000000      4.000000      4.000000  
max            8.000000      9.000000     20.000000

In [7]:
#Representing Categorical Features
df.describe(include=[object])

Month        VisitorType
count   12330              12330
unique     10                  3
top       May  Returning_Visitor
freq     3364              10551

In [8]:
#checking if data contains Null values
df.isnull().sum().sum() 

0

## Feature Extraction

Looking at the columns above, there are 7 categorical columns and 10 numerical columns. 

1) Categorical columns can be scaled only after encoding :
      
      - Weekend and Revenue cannot be label encoded.
      
      - Only month and visitor type can be label encoded.
      

2) Numerical columns can be scaled without any encoding needed.

In [9]:
def ordinal_encode(df, column, ordering):
    df = df.copy()
    df[column] = df[column].apply(lambda x : ordering.index(x))
    return df

def onehot_encode(df, column, prefix):
    df= df.copy();
    dummies = pd.get_dummies(df[column], prefix=prefix)
    df = pd.concat([df, dummies], axis=1);
    df = df.drop(column, axis = 1)
    return df;

In [10]:
month_ordering = ["Jan", "Feb",  "Mar", 
    "Apr", "May", "June", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]
visitor_prefix = "V"

In [11]:
df = ordinal_encode(df, "Month", month_ordering)
df = onehot_encode(df, "VisitorType", visitor_prefix)

df["Weekend"] = df["Weekend"].astype(int)
df["Revenue"] = df["Revenue"].astype(int)

# Training and Testing

In [12]:
#Spliting and Scaling
y = df["Revenue"].copy()
x = df.drop("Revenue", axis = 1).copy()

In [13]:
# Convert categorical variables to numerical using one-hot encoding
data = pd.get_dummies(df)

In [14]:
# Split the dataset into features (X) and target variable (y)
X = data.drop('Revenue', axis=1)
y = data['Revenue'].astype(int)

In [15]:
# Review the X variable DataFrame
X.head()

Administrative  Administrative_Duration  Informational  \
0               0                      0.0              0   
1               0                      0.0              0   
2               0                      0.0              0   
3               0                      0.0              0   
4               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0               1                 0.000000   
1                     0.0               2                64.000000   
2                     0.0               1                 0.000000   
3                     0.0               2                 2.666667   
4                     0.0              10               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay  Month  OperatingSystems  \
0         0.20       0.20         0.0         0.0      1                 1   
1         0.00       0.10         0.0         0.0      1                 2   
2         0.20       0.20         0.0         0.0      1                 4   
3         0.05       0.14         0.0         0.0      1                 3   
4         0.02       0.05         0.0         0.0      1                 3   

   Region  TrafficType  Weekend  V_New_Visitor  V_Other  V_Returning_Visitor  
0       1            1        0              0        0                    1  
1       1            2        0              0        0                    1  
2       9            3        0              0        0                    1  
3       2            4        0              0        0                    1  
4       1            4        1              0        0                    1

In [16]:
# Review the y variable Series
y.head()

0    0
1    0
2    0
3    0
4    0
Name: Revenue, dtype: int32

In [17]:
scaler = StandardScaler()
x = scaler.fit_transform(x)

In [18]:
x_train, x_test, y_train, y_test = train_test_split(x, y, 
                                                    train_size = 0.7, 
                                                    random_state = 20)

x_train, x_test, y_train, y_test = train_test_split(x, y, 
                                                    train_size = 0.7, 
                                                    random_state = 20)

## K-Nearest Neighbors Model

In [19]:
#K-Nearest Neighbors
knn = KNeighborsClassifier(n_neighbors = 9)
knn.fit(x_train, y_train)
knnScore = knn.score(x_test, y_test)
y_pred_knn = knn.predict(x_test)

# Cross Validation Score
knnValidationList = cross_val_score(knn, x, y, cv=5, scoring='accuracy')
knnValidation = sum(knnValidationList) / 5
print (f' Knn Accuracy = {knnScore}')
print (f' Knn Validation = {knnValidation}')

 Knn Accuracy = 0.881859962151933
 Knn Validation = 0.8760746147607461


In [20]:
#K-Nearest Neighbors
knn_clf = KNeighborsClassifier(n_neighbors = 9)
model = knn_clf.fit(x_train, y_train)
y_pred_knn = knn_clf.predict(x_test)

# Cross Validation Score
knn_train=model.score(x_train,y_train)
knn_test=accuracy_score(y_test,y_pred_knn)
knn_precision = metrics.precision_score(y_test,y_pred_knn)
knn_recall = metrics.recall_score(y_test,y_pred_knn)
knn_f1 = metrics.f1_score(y_test,y_pred_knn)
knn_roc = metrics.roc_auc_score(y_test,y_pred_knn)


In [21]:
print(classification_report(y_test,y_pred_knn))

              precision    recall  f1-score   support

           0       0.90      0.97      0.93      3146
           1       0.71      0.35      0.47       553

    accuracy                           0.88      3699
   macro avg       0.80      0.66      0.70      3699
weighted avg       0.87      0.88      0.86      3699



## Naive Bayes Model

In [22]:
#Naive Bayes
nb = GaussianNB()
nb.fit(x_train, y_train)
nbScore = nb.score(x_test, y_test)

# Cross Validation Score
nbValidationList = cross_val_score(nb, x, y, cv=5, scoring='accuracy')
nbValidation = sum(nbValidationList) / 5

print (f' NB Accuracy = {nbScore}')
print (f' NB Validation = {nbValidation}')

 NB Accuracy = 0.7629088942957556
 NB Validation = 0.770235198702352


In [23]:
#Naive Bayes
nb = GaussianNB()
nb.fit(x_train,y_train)
y_pred_nb = nb.predict(x_test)
nb_train = nb.score(x_train,y_train)

# Cross Validation Score
nb_test = accuracy_score(y_test,y_pred_nb)
nb_precision = metrics.precision_score(y_test,y_pred_nb)
nb_recall = metrics.recall_score(y_test,y_pred_nb)
nb_f1 = metrics.f1_score(y_test,y_pred_nb)
nb_roc = metrics.roc_auc_score(y_test,y_pred_nb)

In [24]:
print(classification_report(y_test,y_pred_nb))

              precision    recall  f1-score   support

           0       0.93      0.78      0.85      3146
           1       0.35      0.67      0.46       553

    accuracy                           0.76      3699
   macro avg       0.64      0.72      0.65      3699
weighted avg       0.84      0.76      0.79      3699



## Support Vector Machine Model

In [25]:
#Vector Machine
svm = SVC()
svm.fit(x_train, y_train)
svmScore = svm.score(x_test, y_test)

# Cross Validation Score
svmValidationList = cross_val_score(svm, x, y, cv=5, scoring='accuracy')
svmValidation = sum(svmValidationList) / 5

print (f' SVM Accuracy = {svmScore}')
print (f' SVM Validation = {svmValidation}')

 SVM Accuracy = 0.8961881589618816
 SVM Validation = 0.8892944038929441


In [26]:
#Vector Machine
svm = SVC()
svm.fit(x_train, y_train)
y_pred_svm = svm.predict(x_test)

# Cross Validation Score
svm_train = svm.score(x_train,y_train)
svm_test = accuracy_score(y_test,y_pred_svm)
svm_precision = metrics.precision_score(y_test,y_pred_svm)
svm_recall = metrics.recall_score(y_test,y_pred_svm)
svm_f1 = metrics.f1_score(y_test,y_pred_svm)
svm_roc = metrics.roc_auc_score(y_test,y_pred_svm)


In [27]:
print(classification_report(y_test,y_pred_svm))

              precision    recall  f1-score   support

           0       0.91      0.97      0.94      3146
           1       0.75      0.46      0.57       553

    accuracy                           0.90      3699
   macro avg       0.83      0.72      0.76      3699
weighted avg       0.89      0.90      0.89      3699



## Analysing accuracy and validation

In [28]:
#Analysing accuracy and validation
print("KNN Accuracy: " + str(knnScore))
print("KNN Validation: " + str(knnValidation))
print("------------")

print("NB Accuracy: " + str(nbScore))
print("NB Validation: " + str(nbValidation))

print("------------")
print("SVM Accuracy: " + str(svmScore))
print("SVM Validation: " + str(svmValidation))

KNN Accuracy: 0.881859962151933
KNN Validation: 0.8760746147607461
------------
NB Accuracy: 0.7629088942957556
NB Validation: 0.770235198702352
------------
SVM Accuracy: 0.8961881589618816
SVM Validation: 0.8892944038929441


In [29]:
analysis = {'Training accuracy':[knn_train, nb_train, svm_train],
           'Precision':[knn_precision, nb_precision, svm_precision],
           'Recall':[knn_recall, nb_recall, svm_recall],
           'Test_score':[knn_test, nb_test, svm_test],
            'f1_score':[knn_f1,nb_f1,svm_f1],
            'roc_curve':[knn_roc,nb_roc,svm_roc]
           }

In [30]:
analysis_df = pd.DataFrame(analysis, index = ['KNN','NaiveBayes','SVM'])
analysis_df

Training accuracy  Precision    Recall  Test_score  f1_score  \
KNN                  0.889468   0.710145  0.354430    0.881860  0.472859   
NaiveBayes           0.755648   0.347170  0.665461    0.762909  0.456293   
SVM                  0.899548   0.749263  0.459313    0.896188  0.569507   

            roc_curve  
KNN          0.664501  
NaiveBayes   0.722750  
SVM          0.716147

  - Through metrices we have identified that SVM is the best model with an accuracy of 0.90
  - Precision and Test_score are higher for SVM
  - f1_score is higher for SVM
  - NaiveBayes has more area under ROC curve with 0.72